src/jit/codegenxarch.cpp

```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->info.retStructDesc.passedInRegisters)
+            {
+                assert(compiler->info.retStructDesc.eightByteCount > 0 &&
+                       compiler->info.retStructDesc.eightByteCount <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+                for (unsigned i = 0; i < compiler->info.retStructDesc.eightByteCount; i++)
+                {
+                    var_types eightByteType = compiler->getEightByteType(compiler->info.retStructDesc, i);
+
+                    if (eightByteType == TYP_REF)
+                    {
+                        gcInfo.gcRegGCrefSetCur |= ((i == 0) ? RBM_INTRET : RBM_INTRET_1);
+                    }
+                    else if(eightByteType == TYP_BYREF)
+                    {
+                        gcInfo.gcRegByrefSetCur |= ((i == 0) ? RBM_INTRET : RBM_INTRET_1);
+                    }
+                }
+            }
+        }
+        else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+        {
+            if (compiler->info.compRetType == TYP_REF)
```

Comment:
This should be written this way:

if  (varTypeIsGC(compiler->info.compRetType))
{
    gcInfo.gcRegGCrefSetCur |= RBM_INTRET; 
}